devicetypes/smartthings/viconics-schneider-room-controller.src/viconics-schneider-room-controller.groovy

```diff
@@ -0,0 +1,554 @@
+private getOCCUPANCY_OCCUPIED() { 0x00 }
+private getOCCUPANCY_UNOCCUPIED() { 0x01 }
+private getFAN_SPEED_LOW() { 0x00 }
+private getFAN_SPEED_MEDIUM() { 0x01 }
+private getFAN_SPEED_HIGH() { 0x02 }
+private getFAN_SPEED_AUTO() { 0x03 }
+
+
+private getFAN_MODE_MAP() {
+	[
+		(CUSTOM_FAN_MODE_ON):"on",
+		(CUSTOM_FAN_MODE_AUTO):"auto",
+		(CUSTOM_FAN_MODE_CIRCULATE):"circulate"
+	]
+}
+
+private getTHERMOSTAT_FAN_MODE_ATTRIBUTE_ID_MAP() {
+	[
+		"on":  (CUSTOM_FAN_MODE_ON),
+		"auto": (CUSTOM_FAN_MODE_AUTO),
+		"circulate": (CUSTOM_FAN_MODE_CIRCULATE)
+	]
+}
+
+private getTHERMOSTAT_MODE_MAP() {
+	[
+		(THERMOSTAT_MODE_OFF):"off",
+		(THERMOSTAT_MODE_AUTO):"auto",
+		(THERMOSTAT_MODE_COOL):"cool",
+		(THERMOSTAT_MODE_HEAT):"heat",
+		"04":"heat"
+	]
+}
+
+private getTHERMOSTAT_OPERATING_STATE_MAP() {
+	[
+		(OPERATING_STATE_IDLE):"idle",
+		(OPERATING_STATE_COOLING):"cooling",
+		(OPERATING_STATE_HEATING):"heating"
+	]
+}
+
+private getEFFECTIVE_OCCUPANCY_MAP() {
+	[
+		(OCCUPANCY_OCCUPIED):"Occupied",
+		(OCCUPANCY_UNOCCUPIED):"Unoccupied"
+	]
+}
+
+private getFAN_SPEED_SLIDER_MAP() {
+	[
+		(FAN_SPEED_LOW): 1,
+		(FAN_SPEED_MEDIUM): 2,
+		(FAN_SPEED_HIGH): 3,
+		(FAN_SPEED_AUTO): 4
+	]
+}
+
+def installed() {
+	log.debug "installed"
+	sendEvent(name: "checkInterval", value: 2 * 60 * 60 + 2 * 60, displayed: false, data: [protocol: "zigbee", hubHardwareId: device.hub.hardwareID])
+
+	if (isViconicsVT8350() || isSchneiderSE8350()) {
+		state.supportedFanModes = ["on", "auto"]
+	} else {
+		state.supportedFanModes = ["on", "auto", "circulate"]
+	}
+	state.supportedThermostatModes = ["off", "auto", "cool", "heat"]
+
+	sendEvent(name: "supportedThermostatFanModes", value: JsonOutput.toJson(state.supportedFanModes), displayed: false)
+	sendEvent(name: "supportedThermostatModes", value: JsonOutput.toJson(state.supportedThermostatModes), displayed: false)
+	sendEvent(name: "coolingSetpointRange", value: coolingSetpointRange, displayed: false)
+	sendEvent(name: "heatingSetpointRange", value: heatingSetpointRange, displayed: false)
+}
+
+private void createChildThermostat() {
+	log.debug "Creating child thermostat to handle unoccupied cooling/heating setpoints"
+	def label = "Unoccupied setpoints"
+	def childName = "${device.displayName} ${label}"
+
+	def child = addChildDevice("Child Thermostat Setpoints", "${device.deviceNetworkId}:1", device.hubId,
+			[completedSetup: true, label: childName, isComponent: true, componentName: "childSetpoints", componentLabel: label]
+	)
+
+	child.sendEvent(name: "coolingSetpoint", value: 20.0, unit: "C")
+	child.sendEvent(name: "heatingSetpoint", value: 21.0, unit: "C")
+	log.debug "child.inspect() ${child}"
+}
+
+def parse(String description) {
+    def result = []
+    def eventMap = [:]
+    def descMap = zigbee.parseDescriptionAsMap(description)
+
+    if (descMap.clusterInt == THERMOSTAT_CLUSTER && descMap.attrId) {
+        def attributeInt = zigbee.convertHexToInt(descMap.attrId)
```

Comment:
just use `descMap.attrInt`

Reply:
hmm... why didn't I use that in the first place

Reply:
That's because in most cases there are no **attrInt** in the map returned by zigbee.parseDescriptionAsMap(description)
so it had to be that way:
def attributeInt = zigbee.convertHexToInt(descMap.attrId)

Reply:
I'm looking at the code for `parseDescriptionAsMap`

```
if (outMap.containsKey("attrId")) {
	outMap["attrInt"] = convertHexToInt(outMap["attrId"])
}
```

If this is what you're seeing, I want to see the case in which you are getting a message that is parsing out to have an `attrId` but not an `attrInt`.

Reply:
It was noticed mostly when OCCUPIED reports were processed, that's why i reverted that change.
When i checked it today on a side,  it turned out it worked fine:

`    def description = "read attr - raw: 230A0A02010A0200001800, dni: 230A, endpoint: 0A, cluster: 0201, size: 10, attrId: 0002, result: success, encoding: 18, value: 00"
    def descMap = zigbee.parseDescriptionAsMap(description)
    log.debug "test descMap.inspect(): ${descMap}"`

That was a strange behavior, because yesterday attInt was null in most cases for a while.
Anyway, i'm reverting those changes for now, but i will keep an eye on it.